airflow/traces/tracer.py

```diff
@@ -0,0 +1,280 @@
+        ti,
+        span_name=None,
+        component=None,
+        child=False,
+        links=None,
+    ):
+        """Start a span from taskinstance."""
+        raise NotImplementedError()
+
+
+class EmptyTrace:
+    """If no Tracer is configured, EmptyTracer is used as a fallback."""
+
+    @classmethod
+    def get_tracer(
+        cls,
+        component: str,
+        trace_id: int | None = None,
+        span_id: int | None = None,
+    ):
+        """Get a tracer using provided node id and trace id."""
+        return cls
+
+    @classmethod
+    def start_span(
+        cls,
+        span_name: str,
+        component: str | None = None,
+        parent_sc=None,
+        span_id=None,
+        links=None,
+        start_time=None,
+    ) -> EmptySpan:
+        """Start a span."""
+        return EMPTY_SPAN
+
+    @classmethod
+    def use_span(cls, span) -> EmptySpan:
+        """Use a span as current."""
+        return EMPTY_SPAN
+
+    @classmethod
+    def get_current_span(self) -> EmptySpan:
+        """Get the current span."""
+        return EMPTY_SPAN
+
+    @classmethod
+    def start_span_from_dagrun(
+        cls,
+        dagrun,
+        span_name=None,
+        service_name=None,
+        component=None,
+        links=None,
+    ) -> EmptySpan:
+        """Start a span from dagrun."""
+        return EMPTY_SPAN
+
+    @classmethod
+    def start_span_from_taskinstance(
+        cls,
+        ti,
+        span_name=None,
+        component=None,
+        child=False,
+        links=None,
+    ) -> EmptySpan:
+        """Start a span from taskinstance."""
+        return EMPTY_SPAN
+
+
+class _Trace(type):
+    factory: Callable
+    instance: Tracer | EmptyTrace | None = None
+
+    def __getattr__(cls, name: str) -> str:
+        if not cls.instance:
+            try:
+                cls.instance = cls.factory()
+            except (socket.gaierror, ImportError) as e:
+                log.error("Could not configure Trace: %s, using EmptyTrace instead.", e)
+                cls.instance = EmptyTrace()
+        return getattr(cls.instance, name)
+
+    def __init__(cls, *args, **kwargs) -> None:
+        super().__init__(cls)
+        if not hasattr(cls.__class__, "factory"):
+            if conf.has_option("traces", "otel_on") and conf.getboolean("traces", "otel_on"):
+                from airflow.traces import otel_tracer
+
+                cls.__class__.factory = otel_tracer.get_otel_tracer
+            else:
+                cls.__class__.factory = EmptyTrace
+
+    @classmethod
+    def get_constant_tags(cls) -> str | None:
+        """Get constant tags to add to all traces."""
+        tags_in_string = conf.get("traces", "tags", fallback=None)
+        if not tags_in_string:
+            return None
+        return tags_in_string
+
+
+if TYPE_CHECKING:
+    Trace: EmptyTrace
+else:
+
+    class Trace(metaclass=_Trace):
+        """Empty class for Trace - we use metaclass to inject the right one."""
```

Comment:
Using metaclass here seems a bit unnecessary. I’d prefer having a TraceLike protocol that Trace classes implement, and have (cached) functions that return the correct information as runtime. This would simplify typing as well.

Reply:
I do agree, but I wanted to follow this pattern as otel metrics support followed the same pattern as well. For the future, this might be handy if there is a need to support more than one tracer.